pkg/migration/migrations/public_schema_migration.go

```diff
@@ -0,0 +1,203 @@
+   WHERE ns_db.id != 1
+     AND ns_db."parentSchemaID" = 0
+     AND ns_db."parentID" = 0
+     AND ns_sc."parentSchemaID" = 0
+     AND ns_sc.name = 'public'
+     AND ns_sc.id = 29
+ORDER BY ns_db.id ASC;
+`
+	rows, err := d.InternalExecutor.QueryIterator(
+		ctx, "get_databases_with_synthetic_public_schemas", nil /* txn */, query,
+	)
+	if err != nil {
+		return err
+	}
+	var databaseIDs []descpb.ID
+	for ok, err := rows.Next(ctx); ok; ok, err = rows.Next(ctx) {
+		if err != nil {
+			return err
+		}
+		parentID := descpb.ID(tree.MustBeDInt(rows.Cur()[0]))
+		databaseIDs = append(databaseIDs, parentID)
+	}
+
+	for _, dbID := range databaseIDs {
+		if err := createPublicSchemaForDatabase(ctx, dbID, d); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func createPublicSchemaForDatabase(
+	ctx context.Context, dbID descpb.ID, d migration.TenantDeps,
+) error {
+	return d.CollectionFactory.Txn(ctx, d.InternalExecutor, d.DB, func(
+		ctx context.Context, txn *kv.Txn, descriptors *descs.Collection,
+	) error {
+
+		found, desc, err := descriptors.GetImmutableDatabaseByID(ctx, txn, dbID, tree.DatabaseLookupFlags{})
+		if err != nil {
+			return err
+		}
+		if !found {
+			return errors.Newf("expected to find database with id %d", dbID)
+		}
+		mutableDesc := dbdesc.NewBuilder(desc.DatabaseDesc())
+		dbDesc := mutableDesc.BuildExistingMutable()
+
+		publicSchemaID, err := catalogkv.GenerateUniqueDescID(ctx, d.DB, d.Codec)
+		if err != nil {
+			return err
+		}
+		// Every database must be initialized with the public schema.
+		// Create the SchemaDescriptor.
+		// In postgres, the user "postgres" is the owner of the public schema in a
+		// newly created db. Postgres and Public have USAGE and CREATE privileges.
+		// In CockroachDB, root is our substitute for the postgres user.
+		publicSchemaPrivileges := descpb.NewBasePrivilegeDescriptor(security.AdminRoleName())
+		// By default, everyone has USAGE and CREATE on the public schema.
+		publicSchemaPrivileges.Grant(security.PublicRoleName(), privilege.List{privilege.CREATE, privilege.USAGE})
+		publicSchemaDesc := schemadesc.NewBuilder(&descpb.SchemaDescriptor{
+			ParentID:   dbID,
+			Name:       "public",
+			ID:         publicSchemaID,
+			Privileges: publicSchemaPrivileges,
+			Version:    1,
+		}).BuildCreatedMutableSchema()
+
+		descID := publicSchemaDesc.GetID()
+		idKey := catalogkeys.MakeSchemaNameKey(d.Codec, dbID, publicSchemaDesc.GetName())
+
+		key := catalogkeys.EncodeNameKey(d.Codec, catalogkeys.NewNameKeyComponents(dbID, keys.RootNamespaceID, tree.PublicSchema))
+
+		b := txn.NewBatch()
+
+		// Remove namespace entry for old public schema.
+		b.Del(key)
+		b.CPut(idKey, descID, nil)
+		if err := catalogkv.WriteNewDescToBatch(
+			ctx,
+			false,
+			d.Settings,
+			b,
+			d.Codec,
+			descID,
+			publicSchemaDesc,
+		); err != nil {
+			return err
+		}
+
+		dbDesc.DescriptorProto().GetDatabase().Schemas = map[string]descpb.DatabaseDescriptor_SchemaInfo{
```

Comment:
nit: because `dbDesc` is mutable, you can actually do `dbDesc.Schemas = ...`. If you'd rather not access the embedded struct's field directly, `dbDesc.DatabaseDescriptor().Schemas = ...` is the way to go.